tools/c7n_gcp/c7n_gcp/actions/iampolicy.py

```
@@ -0,0 +1,58 @@
+# Copyright 2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from c7n.utils import type_schema
+from c7n_gcp.actions import MethodAction
+
+
+class SetIamPolicyBaseAction(MethodAction):
+
+    schema = type_schema('setIamPolicy',
+                         required=['bindings'],
+                         **{
+                             'type': {'enum': ['set-iam-policy']},
+                             'bindings': {
+                                 'type': 'array',
+                                 'items': {'role': {'type': 'string'}, 'members': {'type': 'array'}}
+                             }
+                         }
+                         )
+    method_spec = {'op': 'setIamPolicy'}
+
+    MEMBER_TYPES = ['user', 'group', 'domain', 'serviceAccount']
+
+    def get_resource_params(self, model, resource):
+        result = {'resource': resource['name'],
+                  'body': {
+                      'policy': {
+                          'bindings': []
+                      }}
+                  }
+        bindings = result['body']['policy']['bindings']
+
+        if self.data['bindings'] is not None:
+            for binding in self.data['bindings']:
+                if binding['role'] and binding['members']:
+                    members = []
+                    for member in binding['members']:
```

what exactly is going on here with the member type replacing stuff? and why is it needed?

---

there can be incorrect type or missed type. In this case 'user:' will be added automatically

---

what do you mean by incorrect type? Incorrect as in the user passes an invalid string, and so we prepend some text to it? If that's the case then I think it should just throw an error because in that instance the values are just wrong.

or does the API allow for this other kind or something?

---

This part was deleted. Now it works like in GCP without adding new member types